File: lintcode/implement-queue-by-two-stacks.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/implement-queue-by-two-stacks/

class Queue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, element):
        # write your code here
        self.stack1.append(element)

    def top(self):
        # write your code here
        # return the top element
        if self.stack2:
            return self.stack2[-1]
        else:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
            if self.stack2:
                return self.stack2[-1]
        return 0

    def pop(self):
        # write your code here
        # pop and return the top element
        if self.stack2:
            return self.stack2.pop()
        else:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
            if self.stack2:
                return self.stack2.pop()
        return 0

if __name__ == '__main__':
    q = Queue()
    q.push(1)
    print q.pop()
    q.push(2)
    q.push(3)
    print q.top()
    print q.pop()

File: lintcode/implement-queue-by-two-stacks.cc
// Question: http://www.lintcode.com/en/problem/implement-queue-by-two-stacks/
#include <stack>
#include <iostream>
using namespace std;
class Queue {
public:
    stack<int> stack1;
    stack<int> stack2;

    Queue() {
        // do intialization if necessary
    }

    void push(int element) {
        // write your code here
        this->stack1.push(element);
    }

    int pop() {
        // write your code here
        if (!this->stack2.empty()) {
            int res = this->stack2.top();
            this->stack2.pop();
            return res;
        } else {
            while (!this->stack1.empty()) {
                this->stack2.push(this->stack1.top());
                this->stack1.pop();
            }
            if (!this->stack2.empty()) {
                int res = this->stack2.top();
                this->stack2.pop();
                return res;
            }
        }
        return 0;
    }

    int top() {
        // write your code here
        if (!this->stack2.empty()) {
            return this->stack2.top();
        } else {
            while (!this->stack1.empty()) {
                this->stack2.push(this->stack1.top());
                this->stack1.pop();
            }
            if (!this->stack2.empty()) {
                return this->stack2.top();
            }
        }
        return 0;
    }
};

int main(int argc, char** argv) {
    Queue q;
    q.push(1);
    std::cout << q.top() << std::endl;
    q.pop();
    q.push(2);
    q.push(3);
    std::cout << q.top() << std::endl;
    std::cout << q.pop() << std::endl;

    return 0;
}
